src/rpc/blockchain.cpp

```diff
@@ -19,12 +19,14 @@
 #include <hash.h>
 #include <index/blockfilterindex.h>
 #include <index/coinstatsindex.h>
+#include <index/txindex.h>
```

Comment:
Can't you read the scriptpubkey from the undo data?

Reply:
Isn't undo data only kept for recent blocks in the case of reorgs? Didn't check but I can hardly imagine that we store every scriptPubKey of a spent output twice forever.

Reply:
I didn't know about undo data, I pushed a version using it instead of `txindex`. Can revert if undo data is not reliable :/

Reply:
Seems my suspicion was wrong and we indeed keep undo data forever on a non-pruned node. Sorry for confusing!

Reply:
> EDIT: I looked it up and undo data doesn't seem to ever be pruned so I think this is safe?

It does indeed get pruned at the same time as block files https://github.com/bitcoin/bitcoin/blob/master/src/node/blockstorage.cpp#L557-L565.

Reply:
> It does indeed get pruned at the same time as block files [master/src/node/blockstorage.cpp#L557-L565](https://github.com/bitcoin/bitcoin/blob/master/src/node/blockstorage.cpp?rgh-link-date=2022-10-19T19%3A54%3A57Z#L557-L565).

I meant that they are not pruned on non-pruned nodes but you are correct. I've added a check that throws if both pruning and filter_false_positives modes are enabled.

Reply:
But you already added the `GetBlockChecked` and `GetUndoChecked` functions which check and throw if pruned. I don't think it's necessary to disable completely if pruned, since the blocks that are found might not have been pruned yet.

Reply:
Alright, reverted to 134caa7c25fee8ed16d0c390cf729f3033d06dc4.